# Music Streaming Analysis

# Content <a id='back'></a>

* [Introduction](#intro)
* [Step 1. Data Overview](#data_review)
    * [Conclusions](#data_review_conclusions)
* [Step 2. Data Preprocessing](#data_preprocessing)
    * [2.1 Header Style](#header_style)
    * [2.2 Missing Values](#missing_values)
    * [2.3 Duplicates](#duplicates)
    * [2.4 Conclusions](#data_preprocessing_conclusions)
* [Step 3. Hypothesis Testing](#hypothesis)
    * [3.1 Hypothesis 1: User Activity in Both Cities](#activity)
* [Conclusions](#end)

## Introduction <a id='intro'></a>
An analyst's job is to analyze data to gain valuable insights and make informed decisions based on them. This process consists of several steps, such as data overview, data preprocessing, and hypothesis testing.

Whenever we conduct research, we need to formulate a hypothesis that we can later test. Sometimes we accept these hypotheses; other times, we reject them. To make the right choices, a business must be able to understand if it is making the right assumptions or not.

In this project, you will compare the music preferences of the inhabitants of Springfield and Shelbyville. You will study data from an online music streaming service to test the hypothesis presented below and compare the behavior of users in these two cities.

### Objective:
Test the hypothesis:

1. User activity differs depending on the day of the week and the city.


### Steps
Data on user behavior is stored in the file `/datasets/music_project_en.csv`. There is no information about the quality of the data, so it will be necessary to examine it before testing the hypothesis.

First, you will evaluate the quality of the data and see if its problems are significant. Then, during data preprocessing, you will try to address the most critical issues.

Your project will consist of three steps:
1. Data overview
2. Data preprocessing
3. Hypothesis testing








[Return to index](#back)

## Step 1. Data Overview <a id='data_review'></a>

Open the data and examine it.

In [ ]:
# importing pandas
import pandas as pd


Read the file `music_project_en.csv` from the folder `/datasets/` and save it on the variable `df`:

In [ ]:
# reading the file and storing it in df
df = pd.read_csv('/datasets/music_project_en.csv')


In [ ]:
# getting the first 10 rows of the df table
df.head(10)


userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

In [ ]:
# getting general information about our data

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


In [ ]:
# How many duplicate rows are there in the DataFrame
df.duplicated().sum()

3826

In [ ]:
# Number of null values in each column
df.isna().sum()

  userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64

Here are our observations about the table. It contains seven columns. They store the same data type: object.

According to the documentation:

- 'userID' — user identification
- 'Track' — song title
- 'artist' — artist's name
- 'genre' — music genre
- 'City' — user's city
- 'time' — the exact time the song was played
- 'Day' — day of the week

We can see three style issues in the table headers:

1. Some headers are written in uppercase, others are in lowercase.
2. Some headers contain spaces.
3. The 'userID' field has a leading space that could lead to errors later, in addition to having mixed case letters. The other fields have similar issues, where some have mixed case and others are entirely lowercase.

### Write your observations. Here are some questions that can help: <a id='data_review_conclusions'></a>

`1. What kind of data do we have in the rows? And how can we understand the columns?`
All columns have categorical data (object type). Each field can be interpreted as:

'userID' — user identification
'Track' — song title
'artist' — artist's name
'genre' — music genre
'City' — user's city
'time' — the time of day the song was played
'Day' — day of the week

`2. Is this data sufficient to answer our hypothesis, or do we need more data?`
I believe it is possible to answer the proposed hypothesis, but the answer will be quite superficial and with little basis.

One of the main problems I identify is in the 'Day' field, which only informs the name of the day of the week (like "Wednesday"), without providing the full date of playback. This considerably limits the analysis, as valuable information could be obtained from the exact date, such as the season, holidays, and seasonal events—all with the potential to directly influence users' musical behavior.

For example, it is reasonable to assume that on commemorative dates such as June Festivals, Carnival, Christmas, and other regional or national occasions, the most listened to musical styles tend to change. Ignoring this factor can lead to inaccurate or biased interpretations of activity patterns.

Furthermore, the DataFrame has a field that records the exact time the music was played (time). This data should also be considered, as user behavior can vary greatly throughout the day—listening to music at 9 am is not the same as at 11 pm.

`3. Did you notice any problems in the data, such as missing values, duplicates, or wrong data types?`

Yes.

I could see that using the `df.isna().sum()` method revealed the existence of null values in the 'Track', 'artist', and 'genre' columns.

As for duplicate rows, using `df.duplicated().sum()`: 3826.

However, regarding the data types, I believe they are in order, with the exception of the data in the `time` column, which, although I am not very familiar with it, being a string is probably not the most indicated, as there should be a better type when I think about how to use its values.


[Return to index](#back)

## Step 2. Data Preprocessing <a id='data_preprocessing'></a>

The goal here is to prepare the data for analysis.
The first step is to resolve all issues with the header. Then we can move on to missing values and duplicates. Let's begin.

Correct the formatting in the table headers.


### Header Style <a id='header_style'></a>
Print the table headers (the column names):

In [ ]:
# print the column names
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Change the table headers according to good style practices:

- All characters must be lowercase
- Delete spaces
- If the name has multiple words, use snake_case

In [ ]:
# Looping through the headers and converting everything to lowercase
new_names = []

for col in df.columns:
    new_names.append(col.lower())

df.columns = new_names


In [ ]:
# Looping through the headers and removing spaces
new_names = []

for col in df.columns:
    new_names.append(col.strip())

df.columns = new_names

We need to apply the underscore-instead-of-space rule to the `userid` column. It should be `user_id`. Rename this column and print the names of all columns when you're done.

In [ ]:
# Renaming the "userid" column
new_name = {'userid': 'user_id'}
df.rename(columns=new_name, inplace = True)

Check the result. Print the headers again:

In [ ]:
# checking the result: the list of headers
print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Return to index](#back)

### Missing Values <a id='missing_values'></a>

First, find the number of missing values in the table. You need to use two methods in sequence to get the number of missing values.

In [ ]:
# calculating the number of missing values
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

Not all missing values affect the research. For example, the missing values in `track` and `artist` are not critical. You can simply replace them with default values, like the string `unknown`.

But missing values in `genre` can affect the comparison of music preferences in Springfield and Shelbyville. In real life, it would be useful to find out the reasons why the data is missing and try to fix it. But we don't have that possibility in this project. So, you will have to:

- Fill in these missing values with a default value
- Evaluate how much the missing values might affect your analysis

Replace the missing values in the `track`, `artist`, and `genre` columns with the string `unknown`. As we showed in the lessons, the best way to do this is to create a list to store the names of the columns in which we need to make the substitution. Then, use this list and loop through the columns where the substitution is needed and make the substitution.

In [ ]:
# looping through the headers and replacing missing values with 'unknown'
columns = ['track', 'artist', 'genre']
df[columns] = df[columns].fillna('unknown')


Now check the result to make sure the dataset does not contain missing values after the substitution. To do this, count the missing values again.

In [ ]:
# counting the missing values
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Return to index](#back)

### Duplicates <a id='duplicates'></a>
Find the number of explicit duplicates in the table. Remember that you need to apply two methods in sequence to get the number of explicit duplicates.

In [ ]:
# counting explicit duplicates
df.duplicated().sum()


3826

Now discard all duplicates. To do this, call the method that does exactly that.

In [ ]:
# removing explicit duplicates
df = df.drop_duplicates()

Now let's check if we have discarded all duplicates. Count explicit duplicates once more to make sure you have removed all of them:

In [ ]:

# checking for duplicates again
df.duplicated().sum()

0

Now we want to get rid of implicit duplicates in the `genre` column. For example, the name of a genre can be written in different ways. Some errors will also affect the result.

To do this, let's start by printing a list of unique genre names, sorted in alphabetical order. To do this:

- Extract the genre column from the DataFrame
- Call the method that will return all unique values in the extracted column

In [ ]:
# viewing unique genre names
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb', 'hip',
       'jazz', 'postrock', 'latin', 'classical', 'metal', 'reggae',
       'triphop', 'blues', 'instrumental', 'rusrock', 'dnb', 'türk',
       'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock', 

Look at the list and find implicit duplicates of the `hiphop` genre. These could be misspelled names, or alternative names for the same genre.

You will see the following implicit duplicates:

* `hip`
* `hop`
* `hip-hop`

To get rid of them, create a function `replace_wrong_genres()` with two parameters:

* `wrong_genres`= — this is a list containing all the values you need to replace

* `correct_genre`= — this is a string you will use for the substitution

As a result, the function should correct the names in the `genre` column of the `df` table, that is, replacing each value from the `wrong_genres` list with values from `correct_genre`.

Within the function body, use a `for` loop to iterate through the list of wrong genres, extract the `genre` column, and apply the `replace` method to make the corrections.

In [ ]:
# function to replace implicit duplicates
def replace_wrong_genres(data, column, wrong_names, correct_name):
    data[column].replace(wrong_names, correct_name, inplace=True)
    return data

Now, call the `replace_wrong_genres()` function and pass appropriate arguments so that it cleans up implicit duplicates (`hip`, `hop`, and `hip-hop`) by replacing them with `hiphop`:

In [ ]:
# removing implicit duplicates
wrong_names = ['hip', 'hop', 'hip-hop']
correct_name = 'hiphop'
df = replace_wrong_genres(df, 'genre', wrong_names, correct_name)

Make sure the duplicate names have been removed. Print the list of unique values from the `genre` column once more:

In [ ]:
# checking for duplicate values
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb',
       'hiphop', 'jazz', 'postrock', 'latin', 'classical', 'metal',
       'reggae', 'triphop', 'blues', 'instrumental', 'rusrock', 'dnb',
       'türk', 'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock

[Return to index](#back)

### Your observations <a id='data_preprocessing_conclusions'></a>

`Briefly describe what you noticed when analyzing duplicates, as well as the approach you used to eliminate them and the results you achieved.`

When analyzing the data, I realized that in columns with many unique values or similar names, it is common to have implicit duplicates—such as spelling or capitalization variations (`hip-hop`, `HipHop`, `hip`) or even incorrect entries. These records often go unnoticed in a superficial analysis. I believe there are also cases where the analyst is unaware of the correct spelling, and these inconsistencies go unnoticed by those who do not have expertise in the subject.

To eliminate these inconsistencies, I used a manual approach, replacing wrong values with a function that maps the incorrect names to the correct value. Although functional, this approach has limitations: it requires prior knowledge of the existing variations, besides being not very scalable and dependent on manual intervention, which is not attractive.

The result was that I was able to standardize the incorrect records to a single spelling and format for the `hiphop` music genre within the `genre` column.

[Return to index](#back)

## Step 3. Hypothesis Testing
 <a id='hypothesis'></a>

### Hypothesis: comparison of user behavior in the two cities <a id='activity'></a>

The hypothesis states that there are differences in music consumption by users in Springfield and Shelbyville. To test the hypothesis, use the data from the three weekdays: Monday, Wednesday, and Friday.

* Group the users by city.
* Compare the number of songs played by each group on Monday, Wednesday, and Friday.

In [ ]:
df_springfield = df[df['city'] == 'Springfield'] # Springfield Users
df_shelbyville = df[df['city'] == 'Shelbyville'] # Shelbyville Users

In [ ]:
# Counting the songs played in each city
print('Number of songs played in Springfield: ', df_springfield['track'].count())
print('Number of songs played in Shelbyville: ', df_shelbyville['track'].count())

Qtd de músicas tocadas em Springfield:  42741
Qtd de músicas tocadas em Shelbyville:  18512


`Comment on your observations here`

I grouped the user groups from each city into a new variable each.

Now let's group the data by day of the week and find the number of songs played on Monday, Wednesday, and Friday. Use the same approach as before, but now we need to group the data differently.

In [ ]:
# Calculating the songs listened to on each of these three days
def music_quantity_day(data, collumn_day, collumn_music, days):
    music_quntity = [] # list to store the number of songs for each day of the week
    for day in days: # 3 days = monday, wednesday, and friday
        music_quntity.append(data[data[collumn_day] == day][collumn_music].count())
    return music_quntity

# list of days to analyze
days = ['Monday', 'Wednesday', 'Friday']
qtd_springfield = music_quantity_day(df_springfield, 'day', 'track', days)
qtd_shelbyville = music_quantity_day(df_shelbyville, 'day', 'track', days)

for i, day in enumerate(days):
    print(f"{day} - Springfield: {qtd_springfield[i]} / Shelbyville: {qtd_shelbyville[i]}")

Monday - Springfield: 15740 / Shelbyville: 5614
Wednesday - Springfield: 11056 / Shelbyville: 7003
Friday - Springfield: 15945 / Shelbyville: 5895


`Comment on your observations here`

Previously, I had made 6 variations of the print statement, but then I thought it would be less cluttered to use a similar function as I did when handling the inconsistent 'hiphop' data in the 'genre' column.

I made a function that takes a DataFrame, the day column, the music column, and a list with the days for which I want to get the song counts as parameters. It returns a list with the counts for each day, which I save into 2 variables corresponding to each of the two cities, and then I print these values.

You have just learned how to count entries by grouping them by city or by day. And now you need to write a function that can count entries simultaneously based on both criteria.

Create the `number_tracks()` function to calculate the number of songs played on a given day and in a given city. The function should accept two parameters:

* `day`: a day of the week for which we need to filter the data. For example, `Monday`.
* `city`: a city for which we need to filter the data. For example, `Springfield`.

Inside the function, you will apply consecutive filtering with logical indexing.

First, filter the data by day and then filter the resulting table by city.

After filtering the data using both criteria, count the number of values in the 'user_id' column of the resulting table. The result of the count will represent the number of entries you want to find. Store the result in a new variable and print it.

In [ ]:
# Declare the number_tracks() function with two parameters: day= and city=.
def number_tracks(day, city):
    # Filter by day
    filtered_by_day = df[df['day'] == day]
    # Filter by city
    filtered_by_city = filtered_by_day[filtered_by_day['city'] == city]
    # Count how many 'user_id' records exist after the two filters
    total = filtered_by_city['user_id'].count()
    # Return the result
    return total

Call the `number_tracks()` function six times, changing the parameter values, so you can retrieve the data for both cities for each of the three days.

In [ ]:
# the number of songs played in Springfield on Monday
total = number_tracks('Monday', 'Springfield')
print(total)

15740


In [ ]:
# the number of songs played in Shelbyville on Monday
total = number_tracks('Monday', 'Shelbyville')
print(total)

5614


In [ ]:
# the number of songs played in Springfield on Wednesday
total = number_tracks('Wednesday', 'Springfield')
print(total)

11056


In [ ]:
# the number of songs played in Shelbyville on Wednesday
total = number_tracks('Wednesday', 'Shelbyville')
print(total)

7003


In [ ]:
# the number of songs played in Springfield on Friday
total = number_tracks('Friday', 'Springfield')
print(total)

15945


In [ ]:
# the number of songs played in Shelbyville on Friday
total = number_tracks('Friday', 'Shelbyville')
print(total)

5895


**Conclusions**

Comment on whether the third hypothesis is correct or should be rejected. Explain your reasoning.`

As I said before, it is very difficult to get a good result with the current data; there is data that could be added that would likely provide a more accurate answer. Therefore, I believe the hypothesis in its current state should be rejected.

[Return to index](#back)

# Conclusions <a id='end'></a>

Summarize your conclusions about the hypothesis here`

Although the data shows numerical differences in the number of tracks played between Springfield and Shelbyville on certain days of the week, accepting the hypothesis based solely on this would be premature. The DataFrame treats all Mondays, Wednesdays, and Fridays as equivalent, without considering variations between weeks, months, or seasonal events that could influence user behavior. Furthermore, there is no normalization for population size or the number of active users in each city, which makes the comparison arbitrary. That is, the fact that one city has more plays on a given day does not necessarily indicate a distinct behavior pattern—it could simply reflect a larger user base. Therefore, the hypothesis should be rejected or, at the very least, considered inconclusive with the current data.

Simply put, I must reject this hypothesis, as the results provided with the current DataFrame are weak.

[Return to index](#back)